src/protocols/fiat-protocol.js

```diff
@@ -0,0 +1,291 @@
+ * @property {never} [fiatAmount] - The amount of fiat currency to spend, in its smallest unit (e.g., cents for USD).
+ */
+
+/**
+ * @typedef {Object} BuyWithFiatAmountOptions
+ * @property {number | bigint} fiatAmount - The amount of fiat currency to spend, in its smallest unit (e.g., cents for USD).
+ * @property {never} [cryptoAmount] - The amount of crypto asset to buy, in its base unit (e.g., wei for ETH).
+ */
+
+/**
+ * @typedef {SellCommonOptions & (SellExactCryptoAmountOptions | SellForFiatAmountOptions)} SellOptions
+ */
+
+/**
+ * @typedef {Object} SellCommonOptions
+ * @property {string} cryptoAsset - The provider-specific code of the crypto asset to sell.
+ * @property {string} fiatCurrency - The currency's ISO 4217 code (e.g., 'USD').
+ * @property {string} [refundAddress] - The wallet address to receive refunds in case of failure. Defaults to the account's address.
+ */
+
+/**
+ * @typedef {Object} SellExactCryptoAmountOptions
+ * @property {number | bigint} cryptoAmount - The amount of crypto asset to sell, in its base unit (e.g., wei for ETH).
+ * @property {never} [fiatAmount] - The amount of fiat currency to receive, in its smallest unit (e.g., cents for USD).
+ */
+
+/**
+ * @typedef {Object} SellForFiatAmountOptions
+ * @property {number | bigint} fiatAmount - The amount of fiat currency to receive, in its smallest unit (e.g., cents for USD).
+ * @property {never} [cryptoAmount] - The amount of crypto asset to sell, in its base unit (e.g., wei for ETH).
+ */
+
+/**
+ * A protocol-agnostic, standardized object representing a quote for an on/off-ramp transaction.
+ * @typedef {Object} FiatQuote
+ * @property {bigint} cryptoAmount - The amount of the crypto asset, in its base unit (e.g., wei).
+ * @property {bigint} fiatAmount - The amount of the fiat currency, in its smallest unit (e.g., cents).
+ * @property {bigint} fee - The fee charged for the transaction, denominated in the smallest unit of the fiat currency.
+ * @property {string} rate - The effective exchange rate, expressed as a string to avoid precision loss (e.g., a rate of "3000.50" for ETH/USD means 1 ETH = 3000.50 USD). Note: This rate applies to the standard units (e.g., ETH and USD).
+ * @property {Record<string, unknown>} [metadata] - Provider-specific raw data for the quote.
+ */
+
+/**
+ * @interface
+ */
+export class IFiatProtocol {
+  /**
+   * Gets a quote for a crypto asset purchase.
+   * @param {Omit<BuyOptions, 'recipient'>} options - The options for the buy operation.
+   * @returns {Promise<FiatQuote>} A quote for the transaction.
+   */
+  async quoteBuy (options) {
+    throw new NotImplementedError('quoteBuy(options)')
+  }
+
+  /**
+   * Generates a widget URL for a user to purchase a crypto asset with fiat currency.
+   * @param {BuyOptions} options - The options for the buy operation.
+   * @returns {Promise<string>} The URL for the user to complete the purchase.
```

Comment:
I would change the return type of the buy and sell methods to FiatQuote & BuyResult. This would be more consistent with the other wallet and protocol modules, where the return types of the methods always include all the quote fields.

If fetching the quote would make the method run much slower in most fiat protocols, then we can avoid doing so.

```suggestion
   * @returns {Promise<FiatQuote & BuyResult>} The operation's result.
```

Reply:
We can return the quote in buy/sell methods without affecting performance.

Reply:
After reviewing MoonPay again, quoting for sell only supports `cryptoAmount`, not `fiatAmount`. however, `sell` supports both `cryptoAmount` and `fiatAmount`. I recommend not calling quote from `buy` and `sell`, this ensures everything still works, developers can call `quoteSell` and `sell` separately